### PR TITLE
[FLINK-25509][connector-base] Add RecordEvaluator to dynamically stop source based on de-serialized records

### PR DESCRIPTION
**Pull request by ruanhang1993**
## What is the purpose of the change

This pull request add RecordEvaluator to dynamically stop source based on de-serialized records.

## Brief change log

The changes could be found in FLIP-208(https://cwiki.apache.org/confluence/display/FLINK/FLIP-208%3A+Add+RecordEvaluator+to+dynamically+stop+source+based+on+de-serialized+records).

## Verifying this change

This change added unit tests about the new interface.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by ruanhang1993** [comment]
@lindong28 @PatrickRen Could you help to review this PR? Thanks ~

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SingleThreadFetcherManager.java [hunk 1]
```diff
@@ -93,6 +93,14 @@ public void addSplits(List<SplitT> splitsToAdd) {
         }
     }
 
+    @Override
+    public void finishSplits(List<SplitT> finishedSplits) {
```
Comment (lindong28) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SingleThreadFetcherManager.java:97]: Would it be better to name it as `removeSplits()` for consistency with `addSplits()`?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/FinishSplitsTask.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsDeletion;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/** The task to finish reading some splits. */
+public class FinishSplitsTask<SplitT extends SourceSplit> implements SplitFetcherTask {
```
Comment (lindong28) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/FinishSplitsTask.java:32]: Would it be better to name it as `RemoveSplitsTask` for consistency with `AddSplitsTask`?

And should we annotate this class with `@Internal` similar to what we do for `AddSplitsTask`?

**Review by lindong28 — COMMENTED** [review]
Thanks for the PR. Left some comments below.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsDeletion.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.util.List;
+
+/**
+ * A change to delete splits.
+ *
+ * @param <SplitT> the split type.
+ */
+@Experimental
+public class SplitsDeletion<SplitT> extends SplitsChange<SplitT> {
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsDeletion.java:31]: The changes in the kafka connector will raise in another PR after merging this PR. I will refactor the class name.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsDeletion.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.flink.annotation.Experimental;
+
+import java.util.List;
+
+/**
+ * A change to delete splits.
+ *
+ * @param <SplitT> the split type.
+ */
+@Experimental
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsDeletion.java:30]: Yes, the split deletion between the enumerator and reader still exists some problems. I will mark it @Internal.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SingleThreadMultiplexSourceReaderBase.java [hunk 5]
```diff
@@ -113,4 +115,26 @@ public SingleThreadMultiplexSourceReaderBase(
...
+     * This constructor behaves like {@link #SingleThreadMultiplexSourceReaderBase(Supplier,
+     * RecordEmitter, Configuration, SourceReaderContext)}, but accepts a specific {@link
+     * FutureCompletingBlockingQueue}, {@link RecordEvaluator} and {@link
+     * SingleThreadFetcherManager}.
+     */
+    public SingleThreadMultiplexSourceReaderBase(
+            FutureCompletingBlockingQueue<RecordsWithSplitIds<E>> elementsQueue,
+            SingleThreadFetcherManager<E, SplitT> splitFetcherManager,
+            RecordEmitter<E, T, SplitStateT> recordEmitter,
+            @Nullable RecordEvaluator<T> eofRecordEvaluator,
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SingleThreadMultiplexSourceReaderBase.java:129]: The changes in the kafka connector will raise in another PR after merging this PR.  This PR focuses on the module flink-connector-base.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java [hunk 6]
```diff
@@ -328,14 +348,82 @@ private SplitContext(String splitId, SplitStateT state) {
...
+
+        @Override
+        public void markActive() {
+            sourceOutput.markActive();
+        }
+
+        @Override
+        public void collect(T record) {
+            if (!isStreamEnd) {
+                handleEndOfStreamRecord(record);
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java:409]: I forget that the end record will not be sent. I will fix it.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/FinishSplitsTask.java (hunk 2 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/FinishSplitsTask.java:32]: I will fix it.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SingleThreadFetcherManager.java (hunk 1 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SingleThreadFetcherManager.java:97]: I will fix it.

**Comment by ruanhang1993** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java [hunk 7]
```diff
@@ -211,7 +226,12 @@ private boolean moveToNextSplit(
 
         currentSplitContext = splitStates.get(nextSplitId);
         checkState(currentSplitContext != null, "Have records for a split that was not registered");
-        currentSplitOutput = currentSplitContext.getOrCreateSplitOutput(output);
+        currentSplitOutput =
+                currentSplitContext.getOrCreateSplitOutput(
```
Comment (lindong28) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java:230]: Would it be simpler and more readable to pass a `Function<T, Boolean>`  instead of passing `eofRecordEvaluator`, `splitId` and `splitFetcherManager`? Then we won't need to add the extra generic type parameters to `SplitContext`.

I made a commit here [1] to show the resulting code.

[1] https://github.com/apache/flink/commit/215f80fe4640dc6686ca9ebddc6ae42c5fdb4cb3

**Review by lindong28 — COMMENTED** [review]
Thanks for the update. Left some comments below.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SingleThreadMultiplexSourceReaderBase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lindong28) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SingleThreadMultiplexSourceReaderBase.java:129]: Not sure if we have the precedence of splitting a FLIP's implementation into multiple PRs. I am a bit concerned that it can make the FLIP's PR harder to track and manage. It should be OK to split the implementation into multiple commits in the same PR.

Since we can not get concrete user-side benefits before making the changes in the kafka connector etc, would it be simpler to put all of them in the same PR?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableTestBase.java [hunk 8]
```diff
@@ -147,6 +153,46 @@ public void createTestTopic(String topic, int numPartitions, int replicationFact
         }
     }
 
+    public <K, V> void produceToKafka(
```
Comment (lindong28) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableTestBase.java:156]: Should we produce data to Kafka topic similar to the existing tests such as `KafkaTableITCase#testPerPartitionWatermarkWithIdleSource` for consistency?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRecordEvaluator.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package org.apache.flink.streaming.connectors.kafka.testutils;
+
+import org.apache.flink.connector.base.source.reader.RecordEvaluator;
+import org.apache.flink.table.data.RowData;
+
+/** A mock record evaluator. */
+public class MockRecordEvaluator implements RecordEvaluator<RowData> {
```
Comment (lindong28) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRecordEvaluator.java:7]: Would it be simpler to define it as a public class in `KafkaTableITCase`, similar to the existing public class `KafkaTableITCase#TestPartitioner`?

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/mocks/MockRecordEvaluator.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+package org.apache.flink.connector.base.source.reader.mocks;
+
+import org.apache.flink.connector.base.source.reader.RecordEvaluator;
+
+import java.util.function.Function;
+
+/** A mock RecordEvaluator class. */
+public class MockRecordEvaluator implements RecordEvaluator<Integer> {
```
Comment (lindong28) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/mocks/MockRecordEvaluator.java:26]: Would it be simpler to define it as a private class in`SourceReaderBaseTest`, similar to the existing private class SourceReaderBaseTest#OnEventWatermarkGenerator?

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java [hunk 11]
```diff
@@ -486,6 +493,12 @@ protected KafkaSource<RowData> createKafkaSource(
                 break;
         }
 
+        if (recordEvaluator != null) {
+            kafkaSourceBuilder
+                    .setBounded(new NoStoppingOffsetsInitializer())
```
Comment (lindong28) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java:498]: Why do we need this change? I suppose the boundedness of kafka source should be independent of whether `recordEvaluator` is specified.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java [hunk 12]
```diff
@@ -400,6 +407,49 @@ public void testTableSourceSetOffsetResetWithException() {
                                 ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, errorStrategy));
     }
 
+    @Test
+    public void testTableSourceWithRecordEvaluator() {
+        Map<String, String> tableOptions = getBasicSourceOptions();
+        tableOptions.put(
+                "scan.record.evaluator.class",
+                "org.apache.flink.streaming.connectors.kafka.testutils.MockRecordEvaluator");
```
Comment (lindong28) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicTableFactoryTest.java:415]: Would it be better to pass the value as `MockRecordEvaluator.class.getName()` similar to what `testPerPartitionWatermarkKafka` does to get TestPartitioner's class name?

**Review by lindong28 — COMMENTED** [review]
Thanks for the update. Left some comments below.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaSourceReaderTest.java [hunk 13]
```diff
@@ -479,6 +486,125 @@ public void testSupportsPausingOrResumingSplits() throws Exception {
...
+    public void testSupportsUsingRecordEvaluatorWithSplitsFinishedAtMiddle() throws Exception {
+        final Set<String> finishedSplits = new HashSet<>();
+        try (final KafkaSourceReader<Integer> reader =
+                (KafkaSourceReader<Integer>)
+                        createReader(
+                                Boundedness.BOUNDED,
+                                "groupId",
+                                new TestingReaderContext(),
+                                finishedSplits::addAll,
+                                r -> r == 7 || r == NUM_RECORDS_PER_SPLIT + 5)) {
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaSourceReaderTest.java:499]: This test is written before fixing not to send the eof record. I will check these tests and fix them.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaSourceReaderTest.java [hunk 14]
```diff
@@ -479,6 +486,125 @@ public void testSupportsPausingOrResumingSplits() throws Exception {
...
+            KafkaPartitionSplit split2 =
+                    new KafkaPartitionSplit(new TopicPartition(TOPIC, 1), 0, Integer.MAX_VALUE);
+            reader.addSplits(Arrays.asList(split1, split2));
+            reader.notifyNoMoreSplits();
+
+            TestingReaderOutput<Integer> output = new TestingReaderOutput<>();
+            pollUntil(
+                    reader,
+                    output,
+                    () -> output.getEmittedRecords().size() == 14,
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaSourceReaderTest.java:511]: I will change the condition.
But actually the source may consume some more records after the eof record. But these records will not be sent to the sink.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaSourceReaderTest.java [hunk 15]
```diff
@@ -479,6 +486,125 @@ public void testSupportsPausingOrResumingSplits() throws Exception {
...
+                    "The reader cannot get the excepted result before timeout.");
+            InputStatus status;
+            while (true) {
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaSourceReaderTest.java:514]: This loop make sure the source will turn to the right status and the records after the eof record will not be sent.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRecordEvaluator.java (hunk 9 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRecordEvaluator.java:7]: KafkaDynamicTableFactoryTest relies on it too.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaSourceReaderTest.java [hunk 16]
```diff
@@ -479,6 +486,134 @@ public void testSupportsPausingOrResumingSplits() throws Exception {
...
+        final Set<String> finishedSplits = new HashSet<>();
+
+        try (final KafkaSourceReader<Integer> reader =
+                (KafkaSourceReader<Integer>)
+                        createReader(
+                                Boundedness.BOUNDED,
+                                "groupId",
+                                new TestingReaderContext(),
+                                finishedSplits::addAll,
+                                r -> r == 7)) {
```
Comment (lindong28) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaSourceReaderTest.java:591]: It seems that the test is still hard to follow/understand. It is not self-explanatory why users would have the results that meet the `assertThat` conditions specified below after specifying `r -> r == 7` as the eof condition here.

In general we can use variables instead of constant integers to show the the relationship between the user-specified and the expected result, similar how `NUM_RECORDS_PER_SPLIT` is used here [1].

Maybe we can discuss offline regarding the purpose of these tests before updating this PR.

Same for other tests added in this PR.

[1] https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaSourceReaderTest.java#L356

**Review by lindong28 — COMMENTED** [review]
Thanks for the update. Left one comment below.

**Comment by ruanhang1993** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java [hunk 17]
```diff
@@ -153,25 +157,55 @@ public RecordsWithSplitIds<ConsumerRecord<byte[], byte[]>> fetch() throws IOExce
...
+            handleSplitsRemoval(splitsChange);
+        } else {
             throw new UnsupportedOperationException(
                     String.format(
                             "The SplitChange type of %s is not supported.",
                             splitsChange.getClass()));
         }
+    }
+
+    private void handleSplitsRemoval(SplitsChange<KafkaPartitionSplit> splitsRemoval) {
```
Comment (ruibinx) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:195]: Would it be better to use the concrete SplitsRemoval type here to avoid accidentally passing SplitsAddition?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsRemoval.java [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.flink.annotation.Internal;
+
+import java.util.List;
+
+/**
+ * A change to remove splits.
+ *
+ * <p>This type of {@link SplitsChange} is only used when the {@link
+ * org.apache.flink.connector.base.source.reader.RecordEvaluator} reports the end of the stream. The
+ * split removal between the enumerator and reader still exists some problems.
```
Comment (ruibinx) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsRemoval.java:30]: It would be really helpful to me if you could elaborate on what these problems are. Are they some technical issues that are hard to resolve, or are they just left to be done?

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java (hunk 17 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:195]: Judgments have been added to ensure that it is an instance of SplitsRemoval. So I think it is ok.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsRemoval.java (hunk 18 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsRemoval.java:30]: I remember the community have discussed about the split deletion issue. And there may be some problems about the state if we introduce the split deletion. But I can not find that mail. There is no plan on my side for that issue now.

**Comment by MartijnVisser** [comment]
Minor note: we shouldn't add changes to the Kafka connector in the `master` anymore, since it will be externalized after 1.17 is released. Changes to the Kafka connector should be made in https://github.com/apache/flink-connector-kafka

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsRemoval.java (hunk 18 above)
Comment (ruibinx) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsRemoval.java:30]: Thanks, I'll look into it.

**Comment by ruanhang1993** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcher.java [hunk 19]
```diff
@@ -239,6 +242,23 @@ public void addSplits(List<SplitT> splitsToAdd) {
         }
     }
 
+    /**
+     * Notice the split fetcher that some splits finished. This operation is asynchronous.
+     *
+     * @param removedSplits the removed splits.
+     */
+    public void removeSplits(List<SplitT> removedSplits) {
```
Comment (lindong28) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcher.java:250]: nits: removedSplits -> splitsToRemove

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java [hunk 20]
```diff
@@ -153,6 +153,8 @@ public void accept(Throwable t) {
 
     public abstract void addSplits(List<SplitT> splitsToAdd);
 
+    public abstract void removeSplits(List<SplitT> removedSplits);
```
Comment (lindong28) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java:156]: nits: removedSplits -> splitsToRemove

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SingleThreadFetcherManager.java [hunk 21]
```diff
@@ -113,6 +113,14 @@ public void addSplits(List<SplitT> splitsToAdd) {
         }
     }
 
+    @Override
+    public void removeSplits(List<SplitT> removedSplits) {
```
Comment (lindong28) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SingleThreadFetcherManager.java:117]: Would it be more consistent with `addSplits(List<SplitT> splitsToAdd)` to use `splitsToRemove` here?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsRemoval.java (hunk 18 above)
Comment (lindong28) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/splitreader/SplitsRemoval.java:30]: It seems better to remove the statement "between the enumerator and reader still exists some problems" if we can not provide link or detail regarding the problem.

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/SourceReaderBaseTest.java [hunk 22]
```diff
@@ -336,6 +338,53 @@ void testPerSplitWatermark(boolean emitRecordBeforeSplitAddition) throws Excepti
...
+            InputStatus status = reader.pollNext(output);
+            if (status == InputStatus.END_OF_INPUT) {
+                break;
+            }
+            if (status == InputStatus.NOTHING_AVAILABLE) {
+                reader.isAvailable().get();
+            }
+        }
+        List<Integer> excepted =
+                IntStream.concat(IntStream.range(0, split0End), IntStream.range(10, split1End))
```
Comment (lindong28) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/SourceReaderBaseTest.java:381]: Would it be more readable to replace 10 with NUM_RECORDS_PER_SPLIT?

**Review by lindong28 — COMMENTED** [review]
Thanks for the update! LGTM overall. Left some minor comments.

**Comment by lindong28** [comment]
Thanks for the update. LGTM.
